### ISSUE TITLE
Include system-specific library paths in BYTECC/NATIVECC

### ISSUE THREAD

**Issue by whitequark** [ocaml/opam-repository#3093]
I think OCaml compiler recipes on:
- freebsd
- osx/macports
- osx/homebrew
- linux
- others?

should include `-I /usr/local/include -L /usr/local/lib` (different paths on Homebrew, I think).

The reason are commits like https://github.com/dsheets/ocaml-sodium/commit/210a245cb5ca0cfec2723dac5d62f657a6f670c0 and bugs like https://github.com/dsheets/ocaml-sodium/issues/10, which require workarounds in packages for what is a purely system-dependent gimmick which can and should be handled in a single place.

Note that on Linux, Debian generally includes `/usr/local` in search paths, e.g. ocamlfind search path, to enable local overrides, so on Linux this should be also enabled. Perhaps we can just add the `/usr/local` variations on any OS, and Homebrew-specific paths on Homebrew?

/cc @avsm  @AltGr

**Comment by dbuenzli** [comment]
Two years and a half later do you still all think that this should be done ? 

It seems to me that we would be better off solving this through precise `conf-*` packages.
Bulk addition of include directories tends to lead to underspecified packages. I'm rather prefer precise builds.

**Comment by whitequark** [comment]
Yes, I think this should be done, precisely to *avoid* underspecified packages. Right now the `conf-*` packages are updated to look into `/usr/local` on request; the end result is the support matrix is a patchwork. Fundamentally I would consider `/usr/local/lib` a path just like `/lib`, `/usr/lib`, `/usr/lib/x86_64-linux-gnu`, and so on. Imagine if `conf-*` packages had to explicitly know about multilib!

**Comment by dbuenzli** [comment]
I don't mind having an entity that centralizes path lookups for `conf-*` packages. I'm just wondering whether the `ocaml` compiler is the right place to do this.

**Comment by whitequark** [comment]
I have no strong opinion on where this is done, but I do note that if done in the compiler, every package that uses `ocamlfind ocamlc foo.c` will automatically gain support, but if somewhere else, it will have to be manually threaded through a great many build systems, some of which, based on my cross-compilation experience, will need extensive patching to even gain the support for adding search paths.

**Comment by avsm** [comment]
It may be best to have a `conf-cc` package rather than the compiler, as this is something that is fairly C-specific.  I do agree that having a single package is way better for this than scattering it across multiple poorly-specified individual `conf-*` packages.

**Comment by ygrek** [comment]
I always thought this stuff belong to host configuration, i.e. either embedded into compiler configuration distributed by the OS packages or configured in environment by host admin. E.g. on debian now majority of libs live in /usr/lib/x86_64-linux-gnu/ but I never had to think about it for a second because gcc from debian packages knows to look in that place.
Same applies to /usr/local/lib or /opt - once you start installing stuff in non-standard places it is your reponsibility to make it available to gcc et al - and it is as easy as setting two environment variables - `C_INCLUDE_PATH` and `LIBRARY_PATH`.
tl;dr - I believe it is out of scope for opam (it may of course hint users how to configure gcc properly or propose to include such settings into ~/.profile)

**Comment by whitequark** [comment]
Wow, I had no idea those environment variables exist. I agree that it is out of scope of opam then, but should be documented somewhere.

**Comment by dbuenzli** [comment]
In #10196 @whitequark suggested to document this in the opam-depext error message. 

I don't think this is the right place, since the depext might succeed, but *access* to the installed software may not.  I would say that the suggestion to adjust `C_INCLUDE_PATH` and/or `LIBRARY_PATH` should rather be output on `conf-*` package which AFAIU opam v2. distinguishes. 

So if everyone agrees this is the right thing to do I'm going to open an issue on `opam` and close this one.

**Comment by dbuenzli** [comment]
(on osx fundamentally this should be suggested by macports and homebrew themselves).

**Comment by dbuenzli** [comment]
> I would say that the suggestion to adjust C_INCLUDE_PATH and/or LIBRARY_PATH should rather be output on conf-* package which AFAIU opam v2. distinguishes.

In fact it should simply be on the failure of an opam package with `depexts`, currently it suggests to use depext, but it should suggest that if that succeded and the package still fails the variables might need to be adjusted.

**Comment by dbuenzli** [comment]
Since it seems there's agreement that the `opam-repository` is not the place to solve this problem. I'm closing this, feel free to reopen if you disagree. 

I opened https://github.com/ocaml/opam/issues/3049 for `opam` to suggest to adjust these variables on failures with `depexts`.
